DSharpPlus.Core/Rest/RestClient.cs

```diff
@@ -0,0 +1,79 @@
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+using DSharpPlus.Caching.Abstractions;
+using DSharpPlus.Core.Exceptions;
+
+using Microsoft.Extensions.Options;
+
+using Polly;
+using Polly.Wrap;
+
+namespace DSharpPlus.Core.Rest
+{
+    /// <summary>
+    /// Represents a client for all rest requests to Discord.
+    /// </summary>
+    public class RestClient
+    {
+        private readonly ICacheService __cache;
+        private readonly HttpClient __http_client;
+        private readonly AsyncPolicyWrap<HttpResponseMessage>? __wrapped_policy;
+
+        private readonly string __token;
+
+        public RestClient(ICacheService cache, HttpClient client, IOptions<RestClientOptions> options)
+        {
+            __cache = cache;
+            __http_client = client;
+            __token = options.Value.Token;
+
+            __wrapped_policy = Policy.WrapAsync(new PollyRatelimitPolicy(), new PollyRetryPolicy(options.Value.MaximumRetries).RetryPolicy);
+        }
+
+        public async Task<HttpResponseMessage> MakeRequestAsync(IRestRequest request)
+        {
+            Context requestContext = new()
+            {
+                ["cache"] = __cache,
+                ["endpoint"] = request.Endpoint,
+                ["subject-to-global-limit"] = request.IsSubjectToGlobalLimit
+            };
+
+            HttpRequestMessage requestMessage = request.Build();
+
+            if (request.IsSubjectToGlobalLimit)
+            {
+                requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bot", __token);
```

Comment:
Add to default headers instead? Regardless, this will experience the same issue Remora does, because now the RestClient can't be used for OAuth2, which uses the `Bearer` prefix.

Reply:
no, not added to default headers because of how webhook ratelimits work. though this should probably be handled through a separate field on IRestRequest.

Reply:
We're both LGPL so why implement when Jax did the work for you :mmlul:

Reply:
Jax actively encourages us to use Remora as a base when possible. I know you're averse to Results, but still.

Reply:
"averse" is putting it rather lightly